pinot-plugins/pinot-stream-ingestion/pinot-kafka-2.0/src/main/java/org/apache/pinot/plugin/stream/kafka20/KafkaPartitionLevelConsumer.java

```diff
@@ -46,6 +50,20 @@ public KafkaPartitionLevelConsumer(String clientId, StreamConfig streamConfig, i
     super(clientId, streamConfig, partition);
   }
 
+  @Override
+  public void validateStreamState(StreamPartitionMsgOffset startMsgOffset) throws PermanentConsumerException {
+    final long startOffset = ((LongMsgOffset) startMsgOffset).getOffset();
+    Map<TopicPartition, Long> beginningOffsets =
+            _consumer.beginningOffsets(Collections.singletonList(_topicPartition));
+
+    final long beginningOffset = beginningOffsets.getOrDefault(_topicPartition, 0L);
+    if (startOffset < beginningOffset) {
```

Comment:
Is `startOffset > beginningOffset` an acceptable scenario?

Reply:
Yes. That is the common scenario. [KafkaConsumer::position](https://kafka.apache.org/0110/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html#position(org.apache.kafka.common.TopicPartition)) specifies the next record that is somewhere in the middle or end of the stream. Also `startOffset` is usually the same as `KafkaConsumer::position` if the consumer is keeping up.

Reply:
Coming back to this discussion, I realized my confusion: in this case Beginning Offset is the earliest offset in the Kafka topic and Start Offset is where the new Consuming Segment will start reading from.  

I had been thinking in terms of the last offset in the immediately preceding segment.